File: camvideo/web/live-voice.js
(() => {
  const panel = document.createElement('article');
  panel.className = 'panel';
  panel.style.marginBottom = '20px';
  panel.innerHTML = `<div class="panel-heading"><div><span class="eyebrow">APRESENTAÇÃO DO PRODUTO</span><h2>Falas contínuas com IA</h2></div><span id="live-key-state" class="tiktok-status">Verificando chave</span></div>
    <p class="tiktok-note">A OpenAI escreve os roteiros com os dados abaixo. O OmniVoice gera a voz no PC. Cada fala tem aproximadamente um minuto; duas ficam prontas antes da reprodução começar.</p>
    <details class="voice-help"><summary>Configurar chave da OpenAI</summary><form id="live-key-form" class="tiktok-form"><label for="live-key">Chave da API</label><input id="live-key" type="password" autocomplete="off" spellcheck="false" maxlength="512" placeholder="Cole sua chave"><button class="button button-secondary">Salvar chave protegida</button><p class="tiktok-note">Armazenada com proteção do Windows. Os roteiros consomem créditos da sua API; o áudio é local.</p></form></details>
    <form id="live-product-form" class="tiktok-form" style="margin-top:16px"><label for="live-name">Nome do produto</label><input id="live-name" maxlength="120" required placeholder="Nome exato do produto">
    <label for="live-features">Características e informações confirmadas</label><textarea id="live-features" rows="5" maxlength="3000" required placeholder="Materiais, medidas, funções, modo de uso e outras informações verdadeiras"></textarea>
    <label for="live-offer">Oferta e condições (opcional)</label><textarea id="live-offer" rows="2" maxlength="600" placeholder="Preço, desconto e condições, se existirem"></textarea>
    <div class="tiktok-actions"><button class="button button-secondary">Salvar produto</button><button type="button" id="live-prepare" class="button button-secondary">Gerar amostra de 1 minuto</button></div></form>
    <div class="tiktok-form" style="margin-top:16px"><label for="live-quality">Qualidade da geração</label><select id="live-quality"><option value="32">Original — 32 etapas</option><option value="16">Rápida — 16 etapas</option></select>
    <label for="live-minutes">Duração da sessão</label><select id="live-minutes"><option value="15">15 minutos</option><option value="30">30 minutos</option><option value="60" selected>1 hora</option><option value="120">2 horas</option><option value="0">Até eu parar</option></select>
    <p class="tiktok-note">A sessão usa a saída e o volume escolhidos em “Para onde a voz vai?”. Iniciar autoriza novas chamadas à sua API até o fim da sessão. Confira os dados e ouça a amostra antes.</p>
    <div class="tiktok-actions"><button id="live-start" class="button button-primary">Iniciar falas contínuas</button><button id="live-stop" class="button cancel-button" disabled>Parar sessão</button><button id="live-unload" class="button button-secondary">Liberar memória da voz</button></div>
    <p id="live-status" class="tiktok-note" role="status"></p><p id="live-counts" class="tiktok-note"></p>
    <details class="voice-help"><summary>Último roteiro gerado</summary><p id="live-script" style="white-space:pre-wrap"></p></details></div>`;
  $('view-tiktok').querySelector('.voice-studio').prepend(panel);
  const demo = document.createElement('article');
  demo.className = 'panel'; demo.style.marginBottom = '20px';
  demo.innerHTML = `<div class="panel-heading"><div><span class="eyebrow">SIMULAÇÃO · NOMES FICTÍCIOS</span><h2>Teste do sininho de venda</h2></div></div><p class="tiktok-note">Prévia com sininho e agradecimento. A própria voz anuncia que é uma simulação. Clique para gerar e depois ouça no player.</p><button id="sale-demo-generate" class="button button-secondary">Gerar simulação de venda</button><p id="sale-demo-status" class="tiktok-note" role="status">Nenhuma simulação gerada.</p><audio id="sale-demo-audio" controls hidden aria-label="Ouvir simulação de venda"></audio>`;
  panel.after(demo);
  const stockCard = document.createElement('article');
  stockCard.className = 'panel'; stockCard.style.marginBottom = '20px';
  stockCard.innerHTML = `<div class="panel-heading"><div><span class="eyebrow">OFERTA · QUANTIDADE REAL</span><h2>Unidades no valor promocional</h2></div></div><p class="tiktok-note">Informe quantas unidades do produto acima ainda têm o valor promocional. A quantidade entra nos roteiros da amostra e das falas contínuas. Atualize manualmente conforme as vendas; não há baixa automática.</p><form id="stock-form" class="tiktok-form"><label for="stock-remaining">Unidades restantes</label><input id="stock-remaining" type="number" min="0" max="1000000" step="1" required placeholder="Ex.: 12"><div class="tiktok-actions"><button class="button button-primary">Atualizar quantidade nos áudios</button><button id="stock-disable" type="button" class="button button-secondary">Não mencionar estoque</button></div></form><p id="stock-status" class="tiktok-note" role="status"></p><p id="stock-current" class="tiktok-note"></p><p class="tiktok-note">A atualização interrompe a reprodução e renova as falas pendentes; pode haver uma pausa. Zero encerra a sessão. A configuração vale para o nome de produto informado e é reiniciada ao fechar o servidor. Áudios já salvos na biblioteca não são reescritos.</p>`;
  panel.after(stockCard);
  async function saveStock(remaining) {
    try {
      await act('stock_update', {product:$('live-name').value.trim(), remaining});
      $('stock-status').textContent = remaining === null ? 'Menção ao estoque desativada.' : remaining === 0 ? 'Estoque zerado. A sessão foi interrompida; atualize antes de reiniciar.' : 'Quantidade atualizada. Os próximos roteiros usarão esse estoque.';
    } catch(error) { $('stock-status').textContent = error.message; }
  }
  $('stock-form').onsubmit = e => {e.preventDefault(); saveStock(Number($('stock-remaining').value));};
  $('stock-disable').onclick = () => saveStock(null);
  let stockLoaded = false;
  window.addEventListener('voice-state', e => {
    const value = e.detail.live?.stock;
    if (!value) return;
    if (!stockLoaded) {stockLoaded=true; $('stock-remaining').value=value.remaining ?? '';}
    $('stock-current').textContent = value.remaining === null ? 'Sem menção de quantidade nos roteiros.' : `${value.product}: ${value.remaining} unidade(s) no valor promocional.`;
  });
  const sales = document.createElement('article');
  sales.className = 'panel'; sales.style.marginBottom = '20px';
  sales.innerHTML = `<div class="panel-heading"><div><span class="eyebrow">🔔 VENDAS CONFIRMADAS</span><h2>Sininho e agradecimentos</h2></div></div><p class="tiktok-note">Durante as falas contínuas, registre cada compra concluída. O painel sorteia a espera e anuncia a venda entre as falas, com sininho e agradecimento ao nome real informado. Sem pedidos na fila, nenhum aviso é criado.</p><form id="sales-timing-form" class="tiktok-form"><label for="sales-min">Intervalo mínimo (segundos)</label><input id="sales-min" type="number" min="10" max="3600" value="30" required><label for="sales-max">Intervalo máximo (segundos)</label><input id="sales-max" type="number" min="10" max="3600" value="90" required><button class="button button-secondary">Salvar intervalos</button></form><form id="sales-order-form" class="tiktok-form" style="margin-top:16px"><label for="sales-order">Pedidos e nomes reais — um por linha</label><textarea id="sales-order" rows="6" maxlength="7000" required placeholder="PED-1042; Ana&#10;PED-1043; Gabriel&#10;PED-1044; Mariana"></textarea><p class="tiktok-note">Formato: código do pedido; nome real do comprador para o agradecimento. Use somente o nome que pode ser anunciado na live. Envie até 50 linhas por vez e acrescente mais lotes durante a reprodução; até 1000 pendentes.</p><label><input id="sales-confirmed" type="checkbox" required> Confirmo que todas estas compras foram concluídas</label><button id="sales-add" class="button button-primary">Adicionar vendas à fila</button></form><p id="sales-status" class="tiktok-note" role="status"></p><p id="sales-counts" class="tiktok-note"></p><p class="tiktok-note">A espera pode aumentar enquanto uma fala termina. Parar a sessão cancela os avisos pendentes. Pedidos são lembrados apenas nesta execução; não registre novamente uma venda já anunciada.</p>`;
  demo.before(sales);
  $('sales-timing-form').onsubmit = async e => {
    e.preventDefault();
    try { await act('sales_config', {minimum:Number($('sales-min').value), maximum:Number($('sales-max').value)}); $('sales-status').textContent = 'Intervalos salvos para os próximos sorteios.'; }
    catch(error) { $('sales-status').textContent = error.message; }
  };
  let saleSubmitting = false, salesLoaded = false;
  $('sales-order-form').onsubmit = async e => {
    e.preventDefault(); saleSubmitting = true; $('sales-add').disabled = true;
    try {
      const entries = $('sales-order').value.split(/\r?\n/).map(line => line.trim()).filter(Boolean).map(line => { const separator=line.indexOf(';'); return separator < 0 ? {order:line, name:''} : {order:line.slice(0,separator).trim(), name:line.slice(separator+1).trim()}; });
      const registered = await act('sales_add_batch', {entries, confirmed:$('sales-confirmed').checked});
      $('sales-order-form').reset(); $('sales-status').textContent = `${registered.added} vendas adicionadas à fila. Nomes: ${registered.names.join(", ")}. Você pode adicionar outro lote.`;
    } catch(error) { $('sales-status').textContent = error.message; }
    finally { saleSubmitting = false; }
  };
  window.addEventListener('voice-state', e => {
    const state = e.detail.live, values = state?.sales;
    $('sales-add').disabled = saleSubmitting || !state?.active || !state?.continuous;
    if (!values) return;
    if (!salesLoaded) { salesLoaded=true; $('sales-min').value=values.minimum; $('sales-max').value=values.maximum; }
    $('sales-counts').textContent = `${values.queued} pedidos aguardando preparação · ${values.announced} agradecimentos concluídos` + (state.active && state.continuous ? '' : ' · Inicie as falas contínuas para registrar vendas.');
  });
  let demoId = null, demoSubmitting = false;
  $('sale-demo-generate').onclick = async () => {
    demoSubmitting = true; $('sale-demo-generate').disabled = true;
    $('sale-demo-status').textContent = 'Gerando sininho e voz de demonstração...';
    $('sale-demo-audio').hidden = true;
    try { demoId = (await act('sale_demo')).id; }
    catch (error) { $('sale-demo-status').textContent = error.message; }
    finally { demoSubmitting = false; }
  };
  window.addEventListener('voice-state', e => {
    $('sale-demo-generate').disabled = demoSubmitting || e.detail.generation.busy || !!e.detail.live?.active;
    if (!demoId) return;
    const clip = e.detail.clips.find(item => item.id === demoId);
    if (clip) {
      $('sale-demo-audio').src = clip.url; $('sale-demo-audio').hidden = false;
      $('sale-demo-status').textContent = 'Simulação pronta. Aperte reproduzir para ouvir o sininho e o agradecimento fictício.';
      demoId = null;
    } else if (!e.detail.generation.busy) {
      $('sale-demo-status').textContent = e.detail.generation.message; demoId = null;
    }
  });
  let loaded = false;
  const product = () => ({name:$('live-name').value.trim(), features:$('live-features').value.trim(), offer:$('live-offer').value.trim()});
  async function act(action, data={}) {
    const response = await fetch('/api/voice/action', {method:'POST', headers:{'Content-Type':'application/json'}, body:JSON.stringify({action,...data})});
    const result = await response.json();
    if (!response.ok) throw Error(result.error || 'Não foi possível concluir.');
    return result;
  }
  async function run(action, data={}) {
    try { await act(action,data); } catch(error) { toast(error.message); $('live-status').textContent=error.message; }
  }
  $('live-key-form').onsubmit = async e => {
    e.preventDefault();
    const key=$('live-key').value; $('live-key').value='';
    try { await act('save_openai_key',{key}); toast('Chave salva com proteção do Windows'); }
    catch(error) { toast(error.message); }
  };
  $('live-product-form').onsubmit = async e => { e.preventDefault(); try { await act('save_product',{product:product()}); toast('Produto salvo'); } catch(error) { toast(error.message); } };
  function start(action) {
    if (!$('live-product-form').reportValidity()) return;
    run(action,{product:product(), output:$('voice-output').value, volume:Number($('voice-volume').value)/100,
                minutes:Number($('live-minutes').value), steps:Number($('live-quality').value)});
  }
  $('live-start').onclick=()=>start('live_start'); $('live-prepare').onclick=()=>start('live_prepare');
  $('live-stop').onclick=()=>run('live_stop'); $('live-unload').onclick=()=>run('unload');
  window.addEventListener('voice-state', e => {
    const value=e.detail.live;
    if (!value) { $('live-status').textContent='Abra o painel atualizado para usar as falas contínuas.'; return; }
    if (!loaded) { loaded=true; $('live-name').value=value.product.name; $('live-features').value=value.product.features; $('live-offer').value=value.product.offer; }
    $('live-key-state').textContent=value.apiKeyConfigured ? 'Chave configurada' : 'Configure sua chave';
    $('live-start').disabled=$('live-prepare').disabled=value.active || !value.apiKeyConfigured;
    $('live-stop').disabled=!value.active;
    $('live-unload').disabled=value.active || e.detail.generation.busy || e.detail.playback.busy;
    $('live-status').textContent=value.message;
    $('live-status').classList.toggle('voice-error',!!value.error);
    $('live-counts').textContent=`${value.generated} falas geradas · ${value.queued} na fila · ${value.played} reproduzidas · ${value.requests} chamadas OpenAI · ${value.tokens} tokens · ${e.detail.engineLoaded ? 'Motor de voz em memória' : 'Motor de voz descarregado'}`;
    $('live-script').textContent=value.script || 'Nenhum roteiro ainda.';
  });

  const uploads=document.createElement('article'); uploads.className='panel'; uploads.style.marginBottom='20px';
  uploads.innerHTML=`<div class="panel-heading"><div><span class="eyebrow">VÍDEOS DA LIVE</span><h2>Adicionar demonstrações do produto</h2></div></div><p class="tiktok-note">Envie seus vídeos para a biblioteca e selecione um no rascunho. O envio não altera a câmera nem inicia uma transmissão. Use demonstrações e ângulos diferentes; edições não garantem que o TikTok considere o conteúdo original.</p><div class="tiktok-actions"><button class="button button-secondary" id="live-upload-open">Enviar vídeos</button><input type="file" id="live-video-upload" accept="video/*" multiple hidden><button class="button button-secondary" id="live-upload-cancel" disabled>Cancelar envio</button></div><progress id="live-upload-progress" max="100" value="0" style="width:100%;margin-top:14px"></progress><p class="tiktok-note" id="live-upload-status" role="status">Nenhum envio em andamento.</p>`;
  $('view-tiktok').querySelector('#tiktok-draft').closest('article').before(uploads);
  let xhr=null, cancelled=false;
  $('live-upload-open').onclick=()=>$('live-video-upload').click();
  $('live-upload-cancel').onclick=()=>{cancelled=true; xhr?.abort();};
  $('live-video-upload').onchange=async e=>{
    const files=Array.from(e.target.files); cancelled=false;
    $('live-upload-cancel').disabled=false; e.target.disabled=true;
    try {
      for(const file of files) {
        if(cancelled) break;
        const name=await new Promise((resolve,reject)=>{
          xhr=new XMLHttpRequest(); xhr.open('POST','/api/upload'); xhr.setRequestHeader('X-Filename',encodeURIComponent(file.name));
          xhr.upload.onprogress=event=>{if(event.lengthComputable) $('live-upload-progress').value=event.loaded/event.total*100;};
          xhr.onload=()=>{try {const value=JSON.parse(xhr.responseText); if(xhr.status!==200) throw Error(value.error || 'Falha no envio'); resolve(value.name);}catch(error){reject(error);}};
          xhr.onerror=()=>reject(Error('Erro de conexão durante o envio.')); xhr.onabort=()=>reject(Error('Envio cancelado.'));
          $('live-upload-status').textContent=`Enviando ${file.name}...`; xhr.send(file);
        });
        await refresh(); $('tiktok-video').value=name; $('tiktok-video').dispatchEvent(new Event('change'));
        $('live-upload-status').textContent=`${name} adicionado à biblioteca.`;
      }
    }catch(error){$('live-upload-status').textContent=error.message;}
    finally{xhr=null; e.target.value=''; e.target.disabled=false; $('live-upload-cancel').disabled=true;}
  };
})();

File: camvideo/live_voice.py
"""Bounded audio prefetch queue: OpenAI writes, local OmniVoice synthesizes."""
from collections import deque
import atexit
import threading
import time
import random
import re
from voice_manager import read_json


class LiveVoice:
    def __init__(self, voice, writer):
        self.voice, self.writer = voice, writer
        self.lock = threading.RLock()
        self.cancel = threading.Event()
        self.thread = None
        self.sales = deque()
        self.sale_orders = set()
        self.sale_min, self.sale_max = 30, 90
        self.sale_due = float('inf')
        self.sale_announced = 0
        self.last_pseudonym = None
        self.stock = dict(product='', remaining=None, revision=0)
        self.state = dict(active=False, message='Informe o produto e prepare uma amostra.',
                          queued=0, generated=0, played=0, requests=0, tokens=0)
        atexit.register(self.stop)

    def snapshot(self):
        with self.lock:
            return dict(self.state, apiKeyConfigured=self.writer.configured(), product=self.writer.product(), stock=dict(self.stock),
                        sales=dict(queued=len(self.sales), announced=self.sale_announced,
                                   minimum=self.sale_min, maximum=self.sale_max))

    def update_stock(self, product, remaining):
        if not isinstance(product, str) or not 1 <= len(product.strip()) <= 120:
            raise ValueError('Preencha o nome do produto antes de salvar o estoque promocional.')
        if remaining is not None and (type(remaining) is not int or not 0 <= remaining <= 1000000):
            raise ValueError('Informe uma quantidade inteira entre 0 e 1000000.')
        with self.lock:
            self.stock = dict(product=product.strip(), remaining=remaining, revision=self.stock['revision'] + 1)
            if self.state.get('active'):
                if remaining == 0:
                    self.stop()
                    self.state['message'] = 'Estoque promocional esgotado. Parando as falas; atualize antes de reiniciar.'
                else:
                    self.voice.stop('playback')
                    self.state['message'] = 'Estoque atualizado. Renovando as falas com a informação atual.'

    def configure_sales(self, minimum, maximum):
        if type(minimum) is not int or type(maximum) is not int or not 10 <= minimum <= maximum <= 3600:
            raise ValueError('Use intervalos de 10 a 3600 segundos; o máximo deve ser maior ou igual ao mínimo.')
        with self.lock:
            self.sale_min, self.sale_max = minimum, maximum

    def add_sale(self, order, confirmed):
        return dict(pseudonym=self.add_sales([dict(order=order)], confirmed)['pseudonyms'][0])

    def add_sales(self, entries, confirmed):
        if confirmed is not True:
            raise ValueError('Confirme que todas as vendas foram concluídas.')
        if not isinstance(entries, list) or not 1 <= len(entries) <= 50:
            raise ValueError('Adicione de 1 a 50 pedidos por envio. Você pode enviar mais lotes depois.')
        validated, orders = [], set()
        for entry in entries:
            if not isinstance(entry, dict):
                raise ValueError('Pedido inválido.')
            order = entry.get('order')
            real_name = 'name' in entry
            alias = entry.get('name') if real_name else entry.get('alias', '')
            if not isinstance(order, str) or not re.fullmatch(r'[\w-]{1,80}', order):
                raise ValueError('Cada linha precisa do código do pedido: até 80 letras, números ou hífens.')
            if order in orders:
                raise ValueError('Há códigos de pedido repetidos neste lote. Nenhum pedido foi adicionado.')
            if not isinstance(alias, str) or (real_name and not alias.strip()) or len(alias) > 60 or any(not (c.isalpha() or c in " '-") for c in alias):
                raise ValueError('Use nomes de até 60 caracteres, somente letras, espaços, hífens ou apóstrofos.')
            orders.add(order)
            validated.append((order, alias.strip(), real_name))
        with self.lock:
            if not self.state.get('active') or not self.state.get('continuous') or self.cancel.is_set():
                raise ValueError('Inicie as falas contínuas antes de registrar os avisos de venda.')
            if orders & self.sale_orders:
                raise ValueError('Um pedido deste lote já foi registrado. Nenhum pedido foi adicionado.')
            if len(self.sales) + len(entries) > 1000 or len(self.sale_orders) + len(entries) > 10000:
                raise ValueError('Limite de pedidos atingido (1000 pendentes ou 10000 por execução).')
            names = ('Ana Gabriela', 'Mariana', 'Juliana', 'Camila', 'Beatriz',
                     'Rafael', 'Lucas', 'Gabriel', 'Bruno', 'Daniel', 'Clara', 'Helena')
            pseudonyms = []
            for order, alias, real_name in validated:
                pseudonym = alias or random.choice([name for name in names if name != self.last_pseudonym])
                self.last_pseudonym = pseudonym
                speech = (f'{pseudonym}, obrigada pela sua compra e pela confiança!' if real_name else
                          'Compra confirmada! Para preservar a privacidade, '
                          f'vou usar o nome fictício {pseudonym}. Obrigada pela compra e pela confiança, {pseudonym}!')
                self.sales.append(speech)
                self.sale_orders.add(order)
                pseudonyms.append(pseudonym)
            if self.sale_due == float('inf'):
                self.sale_due = time.monotonic() + random.uniform(self.sale_min, self.sale_max)
            return dict(added=len(pseudonyms), names=pseudonyms, pseudonyms=pseudonyms)

    def start(self, product, output='', volume=.8, minutes=60, steps=32, continuous=True):
        with self.lock:
            if self.thread and self.thread.is_alive():
                raise ValueError('Aguarde a sessão anterior parar.')
            if any(self.voice.jobs.values()):
                raise ValueError('Aguarde ou pare a fala atual primeiro.')
            if not self.voice.installed():
                raise ValueError('Instale o OmniVoice primeiro.')
            if not self.writer.configured():
                raise ValueError('Salve sua chave da OpenAI primeiro.')
            if steps not in (16, 32) or minutes not in (0, 15, 30, 60, 120):
                raise ValueError('Configuração da sessão inválida.')
            if not isinstance(volume, (float, int)) or not 0 <= volume <= 1:
                raise ValueError('Volume inválido.')
            if continuous and not any(row['key'] == output for row in self.voice.output_devices):
                raise ValueError('Selecione a saída de áudio da live.')
            product = self.writer.save_product(product)
            if self.stock['product'] == product['name'] and self.stock['remaining'] == 0:
                raise ValueError('Estoque promocional zerado. Atualize a quantidade ou desative a menção antes de iniciar.')
            self.cancel = threading.Event()
            self.state = dict(active=True, message='Preparando roteiro com a OpenAI...', queued=0,
                              generated=0, played=0, requests=0, tokens=0, script='', error=False,
                              continuous=continuous, minutes=minutes, started=time.time())
            self.thread = threading.Thread(target=self._run,
                args=(product, output, volume, minutes, steps, continuous), daemon=True)
            self.thread.start()

    def stop(self):
        with self.lock:
            self.cancel.set()
            if self.state['active']:
                self.voice.stop('generation')
                self.voice.stop('playback')
                self.state['message'] = 'Parando. Uma solicitação à OpenAI já enviada pode terminar.'

    def update(self, **values):
        with self.lock:
            self.state.update(values)

    def _run(self, product, output, volume, minutes, steps, continuous):
        queue, history = deque(), []
        pending = None
        pending_sale = False
        revision = self.stock['revision']
        pending_revision = revision
        sale_clips = set()
        playback = None
        started_playing = False
        deadline = time.monotonic() + minutes * 60 if minutes else float('inf')
        try:
            while not self.cancel.is_set() and time.monotonic() < deadline:
                with self.lock:
                    if revision != self.stock['revision']:
                        revision = self.stock['revision']
                        queue = deque(ident for ident in queue if ident in sale_clips)
                        history.clear()
                        self.update(queued=len(queue))
                with self.voice.lock:
                    generating = self.voice.jobs['generation'] is not None
                    playing = self.voice.jobs['playback'] is not None
                    generation_state = dict(self.voice.states['generation'])
                    playback_state = dict(self.voice.states['playback'])
                if pending and not generating:
                    if not pending_sale and pending_revision != revision:
                        pending = None
                        continue
                    clip = read_json(self.voice.clip_path(pending).with_suffix('.json'), {})
                    if not clip:
                        raise ValueError(generation_state.get('message', 'A voz não foi gerada.'))
                    if pending_sale:
                        queue.appendleft(pending)
                        sale_clips.add(pending)
                        pending_sale = False
                    else:
                        queue.append(pending)
                    pending = None
                    self.update(generated=self.state['generated'] + 1, queued=len(queue))
                    if not continuous:
                        self.update(message='Amostra pronta na biblioteca. Ouça antes de iniciar.', lastClip=queue[0])
                        return
                if playback and not playing:
                    if playback_state.get('error'):
                        raise ValueError(playback_state['message'])
                    self.update(played=self.state['played'] + 1)
                    if playback in sale_clips:
                        sale_clips.remove(playback)
                        with self.lock:
                            self.sale_announced += 1
                            self.sale_due = time.monotonic() + random.uniform(self.sale_min, self.sale_max) if self.sales else float('inf')
                    playback = None
                # Start with two complete files, then synthesize ahead of playback.
                if not playing and queue and (started_playing or len(queue) >= 2):
                    with self.lock:
                        if revision != self.stock['revision'] or self.cancel.is_set():
                            continue
                        playback = queue.popleft()
                        self.voice.play(playback, output, volume)
                    started_playing = True
                    self.update(queued=len(queue), message='Falando sobre o produto e preparando as próximas falas.')
                sale_name = None
                with self.lock:
                    if continuous and not pending and not sale_clips and self.sales and time.monotonic() >= self.sale_due:
                        sale_name = self.sales.popleft()
                        self.sale_due = float('inf')
                if sale_name:
                    pending = self.voice.generate(sale_name, sale_demo=True)
                    pending_sale = True
                    self.update(message='Preparando o agradecimento de uma venda confirmada.')
                elif not pending and len(queue) < 2:
                    with self.lock:
                        pending_revision = self.stock['revision']
                        stock = dict(self.stock) if self.stock['product'] == product['name'] else None
                    self.update(message='Criando a próxima fala com a OpenAI...', requests=self.state['requests'] + 1)
                    text, usage = self.writer.write(product, history, stock=stock)
                    if self.cancel.is_set() or time.monotonic() >= deadline:
                        break
                    if pending_revision != self.stock['revision']:
                        continue
                    history.append(text)
                    history = history[-3:]
                    self.update(script=text, tokens=self.state['tokens'] + usage.get('total_tokens', 0))
                    pending = self.voice.generate(text, steps=steps, target_duration=60,
                        seed=42 + self.state['generated'])
                    self.update(message='Gerando áudio local; a primeira preparação precisa de duas falas.')
                self.cancel.wait(.1)
        except Exception as error:
            self.update(error=True, message=str(error))
        finally:
            if continuous or self.cancel.is_set() or self.state.get('error'):
                self.voice.stop('generation')
                self.voice.stop('playback')
            with self.lock:
                self.state['active'] = False
                self.sales.clear()
                self.sale_due = float('inf')
                if not self.state.get('error') and (continuous or self.cancel.is_set()):
                    self.state['message'] = 'Sessão encerrada. As falas geradas ficaram salvas.'
